File: project-{topic}-enhanced/src/test/java/com/alx/taskmgr/controller/AuthControllerTest.java
```java
package com.alx.taskmgr.controller;

import com.alx.taskmgr.dto.AuthRequest;
import com.alx.taskmgr.exception.UserAlreadyExistsException;
import com.alx.taskmgr.model.Role;
import com.alx.taskmgr.model.User;
import com.alx.taskmgr.repository.UserRepository;
import com.alx.taskmgr.security.JwtService;
import com.alx.taskmgr.security.UserInfo;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.test.web.servlet.MockMvc;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(AuthController.class)
@DisplayName("AuthController Unit Tests")
public class AuthControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @MockBean
    private UserRepository userRepository;

    @MockBean
    private PasswordEncoder passwordEncoder;

    @MockBean
    private JwtService jwtService;

    @MockBean
    private AuthenticationManager authenticationManager;

    @Test
    @DisplayName("Should register a new user successfully")
    void shouldRegisterUserSuccessfully() throws Exception {
        AuthRequest registerRequest = AuthRequest.builder()
                .username("newuser")
                .email("newuser@example.com")
                .password("password123")
                .build();

        when(userRepository.existsByUsername("newuser")).thenReturn(false);
        when(userRepository.existsByEmail("newuser@example.com")).thenReturn(false);
        when(passwordEncoder.encode(any(String.class))).thenReturn("encodedpassword");
        when(userRepository.save(any(User.class))).thenReturn(User.builder().id(1L).username("newuser").build());

        mockMvc.perform(post("/api/auth/register")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(registerRequest)))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$").value("User registered successfully!"));
    }

    @Test
    @DisplayName("Should return conflict if username already exists during registration")
    void shouldReturnConflictWhenUsernameExists() throws Exception {
        AuthRequest registerRequest = AuthRequest.builder()
                .username("existinguser")
                .email("existing@example.com")
                .password("password123")
                .build();

        when(userRepository.existsByUsername("existinguser")).thenReturn(true);

        mockMvc.perform(post("/api/auth/register")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(registerRequest)))
                .andExpect(status().isConflict())
                .andExpect(jsonPath("$.message").value("Username is already taken!"));
    }

    @Test
    @DisplayName("Should return conflict if email already exists during registration")
    void shouldReturnConflictWhenEmailExists() throws Exception {
        AuthRequest registerRequest = AuthRequest.builder()
                .username("newuser")
                .email("existing@example.com")
                .password("password123")
                .build();

        when(userRepository.existsByUsername("newuser")).thenReturn(false);
        when(userRepository.existsByEmail("existing@example.com")).thenReturn(false); // Simulate check before throwing
        when(userRepository.existsByEmail("existing@example.com")).thenThrow(new UserAlreadyExistsException("Email is already in use!")); // Actual behavior

        mockMvc.perform(post("/api/auth/register")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(registerRequest)))
                .andExpect(status().isConflict())
                .andExpect(jsonPath("$.message").value("Email is already in use!"));
    }

    @Test
    @DisplayName("Should login user successfully and return JWT token")
    void shouldLoginUserSuccessfully() throws Exception {
        AuthRequest loginRequest = AuthRequest.builder()
                .username("testuser")
                .password("password123")
                .build();

        Authentication authentication = new UsernamePasswordAuthenticationToken(
                loginRequest.getUsername(), loginRequest.getPassword());

        UserInfo userDetails = UserInfo.builder()
                .id(1L)
                .username("testuser")
                .email("test@example.com")
                .password("encodedpassword")
                .role(Role.ROLE_USER)
                .build();

        when(authenticationManager.authenticate(any(UsernamePasswordAuthenticationToken.class)))
                .thenReturn(authentication);
        when(authentication.getPrincipal()).thenReturn(userDetails);
        when(jwtService.generateToken(any(Authentication.class))).thenReturn("mockJwtToken");

        mockMvc.perform(post("/api/auth/login")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(loginRequest)))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.token").value("mockJwtToken"))
                .andExpect(jsonPath("$.username").value("testuser"))
                .andExpect(jsonPath("$.role").value("ROLE_USER"));
    }

    @Test
    @DisplayName("Should return unauthorized for invalid login credentials")
    void shouldReturnUnauthorizedForInvalidCredentials() throws Exception {
        AuthRequest loginRequest = AuthRequest.builder()
                .username("invaliduser")
                .password("wrongpassword")
                .build();

        when(authenticationManager.authenticate(any(UsernamePasswordAuthenticationToken.class)))
                .thenThrow(new BadCredentialsException("Bad credentials"));

        mockMvc.perform(post("/api/auth/login")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(loginRequest)))
                .andExpect(status().isUnauthorized())
                .andExpect(jsonPath("$.message").value("Bad credentials"));
    }
}
```

File: project-{topic}-advanced/backend/src/tests/controllers/authController.test.ts
import request from 'supertest';
import app from '../../app';
import { AppDataSource } from '../../db/data-source';
import { User, UserRole } from '../../db/entities/User';
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';

jest.mock('bcryptjs');
jest.mock('jsonwebtoken');

describe('Auth Controller Integration Tests', () => {
  let userRepository: any;

  beforeAll(async () => {
    userRepository = AppDataSource.getRepository(User);
  });

  beforeEach(async () => {
    await AppDataSource.getRepository(User).query(`TRUNCATE TABLE "users" RESTART IDENTITY CASCADE;`);
    jest.clearAllMocks();
    (bcrypt.hash as jest.Mock).mockResolvedValue('hashedpassword123');
    (bcrypt.compare as jest.Mock).mockResolvedValue(true);
    (jwt.sign as jest.Mock).mockReturnValue('mock_jwt_token');
  });

  describe('POST /api/auth/register', () => {
    it('should register a new user successfully', async () => {
      const newUser = {
        username: 'testuser',
        email: 'test@example.com',
        password: 'password123',
      };

      const response = await request(app)
        .post('/api/auth/register')
        .send(newUser)
        .expect(201);

      expect(response.body.message).toBe('User registered successfully');
      expect(response.body.user).toHaveProperty('id');
      expect(response.body.user.username).toBe(newUser.username);
      expect(response.body.user.email).toBe(newUser.email);
      expect(response.body.user.role).toBe(UserRole.USER); // Default role
      expect(bcrypt.hash).toHaveBeenCalledWith(newUser.password, 10);

      const userInDb = await userRepository.findOneBy({ email: newUser.email });
      expect(userInDb).toBeDefined();
    });

    it('should return 409 if user already exists', async () => {
      // First, create a user directly in the DB
      const existingUser = userRepository.create({
        username: 'existing',
        email: 'existing@example.com',
        password: 'hashedpassword',
        role: UserRole.USER,
      });
      await userRepository.save(existingUser);

      const response = await request(app)
        .post('/api/auth/register')
        .send({
          username: 'another',
          email: 'existing@example.com', // Duplicate email
          password: 'password123',
        })
        .expect(409);

      expect(response.body.message).toBe('User with that email or username already exists');
    });

    it('should return 400 for missing credentials', async () => {
      const response = await request(app)
        .post('/api/auth/register')
        .send({ username: 'testuser', email: 'test@example.com' }) // Missing password
        .expect(400);

      expect(response.body.message).toBe('Username, email, and password are required');
    });
  });

  describe('POST /api/auth/login', () => {
    let createdUser: User;
    beforeEach(async () => {
      createdUser = userRepository.create({
        username: 'loginuser',
        email: 'login@example.com',
        password: 'hashedpassword123',
        role: UserRole.USER,
      });
      await userRepository.save(createdUser);
    });

    it('should log in a user successfully and return a token', async () => {
      const response = await request(app)
        .post('/api/auth/login')
        .send({ emailOrUsername: 'login@example.com', password: 'password123' })
        .expect(200);

      expect(response.body).toHaveProperty('token', 'mock_jwt_token');
      expect(response.body.user).toHaveProperty('id', createdUser.id);
      expect(response.body.user.email).toBe(createdUser.email);
      expect(bcrypt.compare).toHaveBeenCalledWith('password123', 'hashedpassword123');
      expect(jwt.sign).toHaveBeenCalled();
    });

    it('should return 401 for invalid credentials', async () => {
      (bcrypt.compare as jest.Mock).mockResolvedValue(false); // Simulate wrong password

      const response = await request(app)
        .post('/api/auth/login')
        .send({ emailOrUsername: 'login@example.com', password: 'wrongpassword' })
        .expect(401);

      expect(response.body.message).toBe('Invalid credentials');
    });

    it('should return 400 for missing credentials', async () => {
      const response = await request(app)
        .post('/api/auth/login')
        .send({ emailOrUsername: 'login@example.com' }) // Missing password
        .expect(400);

      expect(response.body.message).toBe('Email/username and password are required');
    });
  });
});